app-cpp/appPtr.hpp

```diff
@@ -0,0 +1,481 @@
+
+    T* operator->();
+    T& operator*();
+    operator bool();
+
+private:
+    T* mData{};
+}; // class unique_ptr<T>
+
+// Inlines
+
+template<typename T>
+inline unique_ptr<T>::unique_ptr(T* inData) : 
+    //Beware: ControlBlock default initializes mStrongCount to 1
+    mData{inData}
+{
+    // Do Nothing
+}
+
+template<typename T>
+inline unique_ptr<T>::~unique_ptr()
+{
+    reset();
+}
+
+template<typename T>
+inline void unique_ptr<T>::reset()
+{
+    // delete is smart enough to do nothing when passed nullptr
+    delete mData;
+    mData = nullptr;
+}
+
+template<typename T>
+inline T* unique_ptr<T>::release()
+{
+    T* data = mData;
+    mData = nullptr;
+    return data;
+}
+
+template<typename T>
+inline unique_ptr<T>::unique_ptr(unique_ptr&& ioMove)
+{
+    reset();
+    mData = ioMove.release();
+}
+
+template<typename T>
+inline unique_ptr<T>& unique_ptr<T>::operator=(unique_ptr&& ioMove)
+{
+    // Moving to self should be a NOP
+    if (this != &ioMove)
+    {
+        reset();
+        mData = ioMove.release();
+    }
+    return *this;
+}
+
+template<typename T>
+inline T* unique_ptr<T>::operator->()
+{
+    return get();
+}
+
+template<typename T>
+inline T& unique_ptr<T>::operator*()
+{
+    return *get();
+}
+
+template<class T> 
+inline unique_ptr<T>::operator bool()
+{
+    return (mData != nullptr);
+}
+
+template<typename T>
+inline T* unique_ptr<T>::get()
+{
+    return mData;
+}
+
+// TRICKY: Variatic template arguments 
+// Template that takes any number and type of arguments
+template<typename T, typename ...Args>
+unique_ptr<T> make_unique(Args... inArgs)
+{
+    T* data = new T{std::forward<Args>(inArgs)...};
+    unique_ptr<T> unique{data};
+    return unique;
+}
+
+#pragma endregion // unique_ptr
+
+//
+/////////////////////////////////////////////////////////////////////////////////
+//
+
+// SHARED POINTER
+template<typename T>
+class shared_ptr
+{
+public:
+    shared_ptr() = default;
+
+    shared_ptr(T* inData) :
+        mData{inData},
+        mControl{new ControlBlock{inData}}
+    {
+        // Note: ControlBlock::Ctor{} sets strong count to 1
+    }
+
+    ~shared_ptr()
+    {
+        reset();
+    }
+
+    void reset()
+    {
+        // Cannot check the count if mControl no longer exists
+        if (mControl != nullptr)
+        {
+            auto [strong, weak] = mControl->DecrementStrong();
+            if (strong <= 0)
+            {
+                if (mData != nullptr)
+                {
+                    delete mData;
+                }
+
+                if (weak <= 0)
+                {
+                    delete mControl;
+                }
+            }
+        }
+        mData = nullptr;
+        mControl = nullptr;
+    }
+
+    // RO5 Methods
+    // Move Ctor
+    shared_ptr(shared_ptr&& ioMove) noexcept
+    {
+        if (ioMove)
+        {
+            std::swap(mData, ioMove.mData);
+            std::swap(mControl, ioMove.mControl);
```

Comment:
You should add a comment about why "move" uses "swap"...
(answer: because "move" also implies "noexcept")